feat: add custom reducers to estimators [DET-3098]

rb-determined-ai opened the pull request:
## Description

Introduce custom reducers for estimator trial.  From the docstring in the PR:

    During distributed evaluation, many types of metrics calculated via ``tf.metrics`` or
    ``tf.keras.metrics`` cannot be aggregated properly from the per-slot final metrics
    calculated by each separate Estimator replica. One example is ``tf.metrics.auc``, where
    the ROC AUC calculated over predictions and labels from a full dataset cannot be derived
    from a set of ROC AUC metrics evaluated over the shards of a dataset. However, with
    ``make_metric``, the ROC AUC could be calculated in distributed training by calling
    ``sklearn.metrics.roc_auc_score`` in a custom ``reducer`` function.

## Test Plan

Lots of manual testing, in addition to adding a new unit test and a new parallel test.

aaron276h commented on harness/determined/estimator/_estimator_context.py:
```diff
@@ -142,3 +144,26 @@ def allgather_metrics(self, metrics: Any) -> List:
+    def _build_allgather_op(self, f: Callable) -> Callable:
+        """Build an op that uses allgather in a way that is safely sequentialized."""
+
+        def decorated(*arg: Any, **kwarg: Any) -> tf.Operation:
```
non-blocking: seems like we are wrapping only internal functions so shouldn't matter that much, but doesn't hurt to add in `functools.wraps()` here.

rb-determined-ai replied:
good idea

aaron276h commented on harness/determined/estimator/_estimator_context.py:
```diff
@@ -142,3 +144,26 @@ def allgather_metrics(self, metrics: Any) -> List:
+    def make_metric(
+        self,
+        metric: Any,
+        reducer: Union[Callable[[List[Any]], Any], "estimator.MetricReducer"],
```
question: why is `estimator.MetricReducer` in quotes?

rb-determined-ai replied:
It is a forward reference, like it is not yet defined when this portion of the code is read by the interpreter.

aaron276h commented on harness/determined/estimator/_reducer.py:
```diff
@@ -0,0 +1,138 @@
+        pass
+
+
+class _SimpleMetricReducer(MetricReducer):
```
This might change after you add in documentation, but my initial impression is that having this SimpleReducer that takes a arbitrary `reduce` function presents so many different options for the API that it becomes confusing. All of this is generally targeted at an advanced use case, and asking them to define their own `MetricReducer` seems ok.

rb-determined-ai replied:
I think they both add a lot of value.  I expounded on the difference in one of the new docstrings I just added.

aaron276h commented on harness/determined/estimator/_reducer.py:
```diff
@@ -0,0 +1,138 @@
+        In the example of the calculating a distributed mean, accumulate might keep a running sum
+        and a count of metrics received:
+
+            def accumulate(self, metric):
```
praise: really nice informative doc strings

rb-determined-ai replied:
thank you!

aaron276h commented on harness/determined/estimator/_estimator_context.py:
```diff
@@ -133,3 +133,98 @@ class EstimatorExperimentalContext(_data_layer.DataLayerContext):
+        Return an estimator-compatible metric which will be calculated properly, even during
+        distributed training.
+
+        Unfortunately, during distributed evaluation, many types of metrics calculated via
```
Drop `Unfortunately`

rb-determined-ai replied:
done

aaron276h commented on harness/determined/estimator/_estimator_context.py:
```diff
@@ -133,3 +133,98 @@ class EstimatorExperimentalContext(_data_layer.DataLayerContext):
+
+        The primary motivation for passing a function as the reducer is simplicity.  Metrics from
+        all batches will be buffered in memory and passed over the network where they will be
+        reduced all at once.  This is probably fine for scalar metrics, or on validation datasets
```
Rephrase: `This is probably fine`

rb-determined-ai replied:
rewrote the whole second half of this paragraph, let me know if you like it.

aaron276h replied:
Yeah i like the current version, left one comment relate to the last paragraph.

aaron276h commented on harness/determined/estimator/_estimator_context.py:
```diff
@@ -142,3 +145,86 @@ def allgather_metrics(self, metrics: Any) -> List:
+                       eval_metric_ops={"my_avg_prediction": my_avg_prediction},
+                   )
+
+        The primary motivation for passing a function as the reducer is simplicity.  Metrics from
```
it would be more clear imo if this came right after the paragraph that describes the two options.

rb-determined-ai replied:
done.

rb-determined-ai commented:
@aaron276h This is now ready for a "for-realsies" review.  The incremental update since the last time you looked is:

 - added unit and parallel e2e test
 - fixed a bug where the `tf.control_dependencies()` was somehow triggering the call to `_DistributedMetric.result()` twice, resulting in building two allgather ops for every metric in the graph.  I fixed this by pre-building every result op and returning the cached op when `result()` is called.
 - the `_allgather_ops` list in the context has to be reset after every `.evaluate()` call, because every `.evaluate()` is going to build a clean graph, and you don't want to call `tf.control_dependencies()` on ops from the old graph.

aaron276h commented on e2e_tests/tests/fixtures/estimator_dataset/model.py:
```diff
@@ -31,14 +31,35 @@
+        return sum(per_slot_metrics)
+
+
+class EstimatorDebugTrial(estimator.EstimatorTrial):
```
non-blocking: can we rename this from Debug to something else

rb-determined-ai replied:
sure, done

aaron276h commented on harness/determined/estimator/_estimator_context.py:
```diff
@@ -142,3 +145,92 @@ def allgather_metrics(self, metrics: Any) -> List:
+        memory usage, and minimizing the network communication before the final
+        ``cross_slot_reduce`` operation.
+
+        In either the single-function reducer case or the ``det.estimator.MetricsReducer`` case,
```
I couldn't really understand this part

rb-determined-ai replied:
Cool, rewrote this for better clarity.

aaron276h commented on harness/determined/estimator/_reducer.py:
```diff
@@ -0,0 +1,161 @@
+from determined import estimator
+
+
+class MetricReducer:
```
Question: do we want to show this in the docs as well?

rb-determined-ai replied:
oops, definitely yes.  Added.

aaron276h commented on harness/tests/experiment/fixtures/estimator_linear_model.py:
```diff
@@ -0,0 +1,111 @@
+        return sum(per_slot_metrics)
+
+
+class LinearEstimator(estimator.EstimatorTrial):
```
praise: nice tests

rb-determined-ai replied:
thank you!

aaron276h commented on e2e_tests/tests/experiment/test_tf_estimator.py:
```diff
@@ -150,6 +150,28 @@ def test_mnist_estimator_data_layer_lfs(tf2: bool) -> None:
+    config = conf.load_config(conf.fixtures_path("estimator_dataset/distributed.yaml"))
+    # Run with multiple steps to verify we are resetting reducers right.
+    config = conf.set_max_steps(config, 2)
+    config = conf.set_slots_per_trial(config, 2)
```
blocking: this experiment will never get scheduled since this needs to be 8 slots.

rb-determined-ai replied:
good catch.  I was testing on 2 GPUs, and I thought I updated it back to 8, but I guess not.

aaron276h commented on e2e_tests/tests/experiment/test_tf_estimator.py:
```diff
@@ -150,6 +150,28 @@ def test_mnist_estimator_data_layer_lfs(tf2: bool) -> None:
+        config, conf.fixtures_path("estimator_dataset"), 1
+    )
+
+    trials = exp.experiment_trials(experiment_id)
```
non-blocking: just make this: `trial = exp.experiment_trials(experiment_id)[0]`

rb-determined-ai replied:
Done.

aaron276h commented on harness/determined/estimator/_reducer.py:
```diff
@@ -0,0 +1,168 @@
+
+        self.update_state(metric)
+
+        @self._det_context._build_allgather_op
```
question: why make this a decorator and not just a function call?

rb-determined-ai replied:
Talked offline, the question was about calling `_build_allgather_op` directly rather than applying it as a decorator.  I don't care either way, so I took the direct call approach.

rb-determined-ai commented on docs/reference/api/estimator.txt:
```diff
@@ -44,7 +44,23 @@ API or Native API.
+Reducing Metrics
+~~~~~~~~~~~~~~~~
+
+Determined supports proper reduction of arbitrary metrics during distributed
```
I made the following edits to the docstrings:
```diff
 Reducing Metrics
 ~~~~~~~~~~~~~~~~

-Determined supports proper reduction of arbitrary metrics during distributed
-training by allowing users to define custom reducers for their metrics. Custom
-reducers can be either a function or an implementation of the
+Determined supports proper reduction of arbitrary validation metrics during
+distributed training by allowing users to define custom reducers for their
+metrics. Custom reducers can be either a function or an implementation of the
```
```diff
     def make_metric(..) -> tf.keras.metrics.Metric:
         """
-        Return an estimator-compatible metric which will be calculated properly, even during
-        distributed training.
+        Return an estimator-compatible validation metric which will be calculated properly, even
+        during distributed evaluation.
```
```diff
 class MetricReducer:
     """
-    Efficiently aggregating metrics across a multi-slot distributed evaluation is done in two steps:
+    Efficiently aggregating validation metrics across a multi-slot distributed evaluation is done
+    in two steps:
```
